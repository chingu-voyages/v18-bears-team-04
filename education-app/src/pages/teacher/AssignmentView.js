import React, { useState, useEffect } from "react";

import config from "../../config";
import moment from "moment";

import ApiService from "../../services/api-services";
import SideNav from "../../components/SideNav";
import ValidationError from "../../components/ValidationError";

import styled from "styled-components";

const AssignmentView = (props) => {
	const [assignmentInfo, setInfo] = useState(null);
	const [fileUrl, setFileUrl] = useState([]);
	const [assignmentId, setAssignmentId] = useState(null);
	const [error, setError] = useState(null);

	const getAssignmentInfo = (props) => {
		ApiService.getAssignments()
			.then((res) => {
				const filterAssignment = (a) =>
					a.assignmentResults.find(
						(b) => b._id === props.match.params.assignmentId
					);

				const filteredAssignment = res.find((a) => filterAssignment(a));
				setAssignmentId(filteredAssignment._id);
				const currentAssignment = filteredAssignment.assignmentResults.find(
					(a) => a._id === props.match.params.assignmentId
				);

				setInfo(currentAssignment);

				let list = [];
				if (currentAssignment.studentDocLink.length > 0) {
					list = currentAssignment.studentDocLink.map(
						(a) => config.FILE_BASE_URL + a
					);
				}
				setFileUrl(list);
			})

			.catch((err) => setError({ error: err }));
	};

	useEffect(() => {
		getAssignmentInfo(props);
	}, [props]);

	const errorMessage = () => {
		if (error != null) {
			return `Something went wrong.`;
		}
	};

	const renderCurrentFiles =
		fileUrl !== null
			? fileUrl.map((a, index) => {
					return (
						<a key={index} className='download-btn' href={a} download>
							Download File {index + 1}
						</a>
					);
			  })
			: null;

	const handleFeedbackChange = (e) => {
		e.preventDefault();
		setInfo({
			...assignmentInfo,
			teacherFeedback: e.target.value,
		});
	};

	const handleSubmitFeedback = (e) => {
		e.preventDefault();

		const obj = {
			assignmentId: assignmentId,
			studentId: assignmentInfo.studentId,
			teacherFeedback: assignmentInfo.teacherFeedback,
		};
		ApiService.addTeacherFeedback(obj)
			.then((res) => props.history.goBack())
			.catch((err) => setError({ error: err }));
	};

	const formatDate = () => {
		if (assignmentInfo.submittedOnDate === null) {
			return;
		}
		return moment(assignmentInfo.submittedOnDate).format("MMMM DD, YYYY");
	};

	return (
		<>
			<SideNav />
			<AssignmentViewStyle>
				<div className='assignment-view-container'>
					<h1>Assignment View</h1>
					<div className='submission-container'>
						{error !== null && <ValidationError message={errorMessage()} />}
						<div className='student-submission-view-container'>
							<h2> Student: {props.match.params.studentUsername}</h2>
							<div className='file-downloads'> {renderCurrentFiles}</div>
							<span className='submitted-date'>
								Submitted on {assignmentInfo !== null && formatDate()}
							</span>
							<div className='submission-text'>
								{assignmentInfo !== null && assignmentInfo.studentAnswers}
							</div>
						</div>

						<div className='feedback-container'>
							<h2>Feedback</h2>
							<form
								htmlFor='teacher-feedback'
								onSubmit={(e) => handleSubmitFeedback(e)}
							>
								<textarea
									className='feedback-textarea'
									value={
										assignmentInfo !== null
											? assignmentInfo.teacherFeedback
											: " "
									}
									onChange={(e) => handleFeedbackChange(e)}
								/>
								<button>Submit</button>
							</form>
						</div>
					</div>
				</div>
			</AssignmentViewStyle>
		</>
	);
};

const AssignmentViewStyle = styled.div`
	.assignment-view-container {
		padding-left: 250px;
		padding-top: 60px;
		width: 90vw;
		height: 100%;
		margin-left: auto;
		margin-right: auto;
	}
	h1 {
		font-size: 4rem;
		color: #00a3ff;
		margin: 20px;
	}
	h2 {
		font-size: 3rem;
		margin: 20px 20px;
		color: #888888;
	}
	.submission-view {
		width: 90%;
		height: 500px;
		margin: 20px;
		border: 1px solid #a9a9a9;
	}
	.student-submission-view-container,
	.feedback-container {
		display: flex;
		flex-direction: column;
		margin: 40px 40px;
		box-shadow: 2px 3px 3px 3px #a9a9a9;
		border-radius: 10px;
		form {
			display: flex;
			flex-direction: column;
		}
	}
	.submission-text {
		margin: 20px;
    	width: auto;
    	display: block;
    	height: 800px;
    	border: 1px solid #a9a9a9;
   		font-size: 2rem;
    	overflow-y: scroll;
 		padding: 10px;
	}
	.download-btn {
		background-color: #00a3ff;
    	line-height: 40px;
    	text-align: center;
    	font-size: 2rem;
    	width: 200px;
    	height: 40px;
    	margin: 3px auto 3px 20px;
    	color: #ffffff;
    	border-radius: 5px;
    	padding: 5px;
	}
	.submitted-date{
		font-size: 1.25rem;
		margin: 3px auto 3px 20px;
	}
	button{
		background-color: #00a3ff;
		font-size: 2rem;
		width: 200px;
		height: 50px;
		color: #ffffff;
		border-radius: 10px;
		margin-left: auto;
		margin-right: auto;
		margin-bottom: 20px;
	}
	button:hover, .download-btn:hover {
		box-shadow: 3px 3px 3px #a9a9a9;
		cursor: pointer;
	}
	.feedback-textarea {
		margin: 20px;
		height: 100px;
        width: auto;
        border: 1px solid #a9a9a9;
        border-radius: 10px;
        padding: 10px;
        font-size: 1.5rem;
}
	}
	
`;

export default AssignmentView;
