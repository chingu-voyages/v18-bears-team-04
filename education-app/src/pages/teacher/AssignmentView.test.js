import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import AssignmentView from "./AssignmentView";

describe("<AssignmentView />", () => {
	it("renders without crashing", () => {
		const match = { params: { userName: "joteacher" } };
		const div = document.createElement("div");
		ReactDOM.render(
			<BrowserRouter>
				<AssignmentView match={match} />
			</BrowserRouter>,
			div
		);
		ReactDOM.unmountComponentAtNode(div);
	});

	it("renders as expected", () => {
		const match = { params: { userName: "joteacher" } };
		const tree = renderer
			.create(
				<BrowserRouter>
					<AssignmentView match={match} />
				</BrowserRouter>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
