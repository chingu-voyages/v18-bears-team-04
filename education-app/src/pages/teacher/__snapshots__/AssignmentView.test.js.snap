// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssignmentView /> renders as expected 1`] = `
Array [
  <div
    className="sc-AxirZ HJAhU"
  >
    <div
      className="wrap"
    >
      <div
        className="userInfo"
      >
        <div
          className="prof-img"
        >
          <img
            alt="something that looks like you"
            onError={[Function]}
            src="defaultImg.png"
          />
        </div>
        <p
          className="user-name"
        />
        <p
          className="user-type"
        />
      </div>
      <div
        className="links"
      />
    </div>
  </div>,
  <div
    className="sc-AxiKw hFZwvH"
  >
    <div
      className="assignment-view-container"
    >
      <h1>
        Assignment View
      </h1>
      <div
        className="submission-container"
      >
        <div
          className="student-submission-view-container"
        >
          <h2>
             Student: 
          </h2>
          <div
            className="file-downloads"
          >
             
          </div>
          <span
            className="submitted-date"
          >
            Submitted on 
          </span>
          <div
            className="submission-text"
          />
        </div>
        <div
          className="feedback-container"
        >
          <h2>
            Feedback
          </h2>
          <form
            htmlFor="teacher-feedback"
            onSubmit={[Function]}
          >
            <textarea
              className="feedback-textarea"
              onChange={[Function]}
              value=" "
            />
            <button>
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;
